File: csharp-playwright-nunit-reqnroll/SauceDemo.UiTests/Pages/CheckoutOverviewPage.cs
using Microsoft.Playwright;

namespace SauceDemo.UiTests.Pages;

public sealed class CheckoutOverviewPage : BasePage
{
    private const string ItemNames = "[data-test='inventory-item-name']";
    private const string TotalLabel = "[data-test='total-label']";
    private const string FinishButton = "[data-test='finish']";

    public CheckoutOverviewPage(IPage page) : base(page)
    {
    }

    public async Task<IReadOnlyList<string>> ListItemNamesAsync() =>
        await Page.Locator(ItemNames).AllTextContentsAsync();

    public Task<string> GetTotalLabelAsync() => TextOfAsync(TotalLabel);

    public async Task<CheckoutCompletePage> FinishAsync()
    {
        await ClickAsync(FinishButton);
        return new CheckoutCompletePage(Page);
    }
}

File: ts-playwright-test/src/pages/checkoutOverviewPage.ts
import type { Locator, Page } from '@playwright/test';
import { BasePage } from './basePage';
import type { TestLogger } from '../support/testLogger';
import { CheckoutCompletePage } from './checkoutCompletePage';

export class CheckoutOverviewPage extends BasePage {
  readonly itemNames: Locator;
  readonly totalLabel: Locator;
  private readonly finishButton: Locator;

  constructor(page: Page, log: TestLogger) {
    super(page, log);
    this.itemNames = page.getByTestId('inventory-item-name');
    this.totalLabel = page.getByTestId('total-label');
    this.finishButton = page.getByTestId('finish');
  }

  async finish(): Promise<CheckoutCompletePage> {
    await this.finishButton.click();
    return new CheckoutCompletePage(this.page, this.log);
  }
}
